From 23d76bf6323564baca6ed27a6a25a59fcb4591e8 Mon Sep 17 00:00:00 2001
From: Ran Shalit <rshalit@habana.ai>
Date: Thu, 10 Sep 2020 11:50:28 +0300
Subject: [PATCH] [SW-12754] gaudi: security: add debug protection

Add debug protection by using nPRESETDBG bit in
mmCPU_CA53_CFG_ARM_RST_CONTROL register in order to
disable debug and jtag functionality. Disabling
debug functionality is done only when dip switch
for debug is in disabled state.

Change-Id: I735d83a46a29c219c939028d4933d81f2bee4863
Signed-off-by: Ran Shalit <rshalit@habana.ai>
---

diff --git a/gaudi/asic.c b/gaudi/asic.c
index 17185c8..413c123 100644
--- a/gaudi/asic.c
+++ b/gaudi/asic.c
@@ -133,3 +133,14 @@
 	regmap_writel(0, mmPCIE_WRAP_LBW_DRAIN_CFG);
 	regmap_writel(0, mmPCIE_WRAP_HBW_DRAIN_CFG);
 }
+
+#define BOOTSTRAP_JTAG_EN BIT(12)
+
+bool hl_is_jtag_enabled(void)
+{
+	u32 boot_strap;
+
+	/* jtag is enabled if switch is off ('0') */
+	boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS);
+	return !(boot_strap & BOOTSTRAP_JTAG_EN);
+}
diff --git a/include/preboot.h b/include/preboot.h
index 7e72eb7..72cd473 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -104,6 +104,7 @@
 void hl_plat_init(void);
 void hl_ts_init(void);
 bool hl_is_debug_mode(enum hl_debug_hw_func func);
+bool hl_is_jtag_enabled(void);
 
 void preboot_pcie_wait_for_link(void);
 void preboot_delay(void);
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index 78994fb..1d6b7f1 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -17,6 +17,7 @@
 #include <spi.h>
 #include "boot_desc.h"
 #include "pll.h"
+#include <boot_if.h>
 
 static enum fw_post_reset_cmd preboot_image_offset(struct boot_desc *desc)
 {
@@ -131,13 +132,23 @@
 	preboot_puts("PCIe link is UP!\n");
 }
 
-static void hl_jtag_init(void)
+__weak bool hl_is_jtag_enabled(void)
 {
-	u32 reg, nPRESETDBG = BIT(12);
+	return true;
+}
 
-	/* Enable JTAG attach */
+void hl_jtag_init(void)
+{
+	u32 reg;
+	bool is_debug_en;
+
+	is_debug_en = hl_is_jtag_enabled();
+	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_DEBUG_EN, is_debug_en);
 	reg = regmap_readl(mmCPU_CA53_CFG_ARM_RST_CONTROL);
-	reg |= nPRESETDBG;
+	if (is_debug_en)
+		reg |= BIT(12);
+	else
+		reg &= ~BIT(12);
 	regmap_writel(reg, mmCPU_CA53_CFG_ARM_RST_CONTROL);
 }
 
